site/bps/BP-24-BookieScanner.md

```diff
@@ -0,0 +1,92 @@
+
+
+The mechanism bookie scanner to decide which ledgers and entry log files to scan is as follows:
+
+
+* When a bookie is serving read requests, if an IOException is caught, then the entry log file is marked as suspicious and added to the scanner’s suspicious entry log list, if a NoSuchLedger or NoSuchEntry exception is caught, then the given ledger is marked as suspicious and added to the scanner’s suspicious ledger list.
+* The bookie scanner loops over all the ledger index files. At each iteration, it checks one ledger.
+   * If the suspicious ledger list is not empty, it pops one suspicious ledger to scan
+* The bookie scanner loops over all the entry log files. At each iteration, it checks one entry log file.
+   * If the suspicious entry log list is not empty, it popos one suspicious entry log to scan
+
+
+
+
+#### Scan Cursor
+
+
+To keep track of the scanning position among the ledgers and entry log files, a cursor is maintained. The cursor is saved to disk periodically (it should be configurable). This way, even the bookie process restarts or the server reboots, the scan doesn’t have to restart from the very beginning.
+
+
+#### Scan Throttling
+
+
+The scanner is I/O consumption. We cannot afford to loop scanning them continuously, because this could create busy I/Os and harm normal I/O performance. Instead, the scanners run at a configured rate for throttling purpose (similar as throttling at compaction), with appropriate sleep intervals between two scan periods. When a ledger or an entry log file is marked as suspicious, the bookie scanner is woken up if it is waiting for the next scan period.
+
+
+Some pseudocode to show the callable example:
+
+
+BookieScanner scanner = new BookieScanner();
+scanner.scanLedgers();// find the missing ledgers, and check whether entries are missing, for condition1 and condition2
+scanner.scanEntryLogs();// check whether the entries in the entry log are corrupted, for condition 3 
+scanner.scanSuspiciousLedgers();
+scanner.scanSuspiciousEntryLogs();
+scanner.fixCorrupt();//replicate lost fragments from replica
+
+
+
+
+### Compatibility, Deprecation, and Migration Plan
+
+
+- This new feature has no impacts to existing users.
+
+
+
+
+### Test Plan
+
+
+After we delete some entry logs which contain some ledgers' data, we will get NoSuchEntry or NoSuchLedgerthe Exception, and the BookieScanner will be triggered to recovery all the data from replica. After a period, we can check that by readEntry from the bookie. Addition, we can check BookieScanner’s ability to recovery index files using similar way.
+
+
+### Rejected Alternatives
+
+
+We can achieve the same stuff by adding one dedicated thread to check the ledger's state and scan the index files and entry logs to verify entry's integrity regularly, but that could lead much high overhead. And the actions needed by this task just part of compacting task, such as scanning the entry log, so we can integrate BookieScanner to it.
```

Comment:
I guess I don't understand fully what you have rejected. Can you elaborate more? Again the scope of this work is large, and very high-level design here may not be enough. I would like to request a google doc with more details and iterate over that.

Reply:
These words are more about where to execute the logic, which is far way the `alternative`. I think the alternative way is enhancing LedgerChcker by checking more elaborately, which has more overhead. I added these in google doc.

Reply:
Hi @ArvinDevel  any update/progress on this ?

Reply:
@jvrao no, sorry for that I have no time to improve it, maybe we can close this one.

Reply:
yeah, here is [my branch](https://github.com/ArvinDevel/bookkeeper/tree/diskScrubber) for this work.
@sijie @jvrao